### PR TITLE
Optimize merge for CO and Taxon

### LINKED ISSUE #4019: Use the same optimization for collection object table as done for spauditlog (agent merging)
Since there is no uniqueness rules or business rules regarding collection object table to agent relationships the same optimization can be done as done for spauditlog here https://github.com/specify/specify7/pull/3855. Would be very beneficial since I've noticed that agent merging spends a lot of time replacing agent ids in collection object table.

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #4019

### CODE REVIEW

**Comment by acwhite211** [comment]
Lol, I created a copy PR https://github.com/specify/specify7/pull/4103.  The only difference is that I edited the comments.  Also, the `table_name.lower() in MERGING_OPTIMIZATION_FIELDS[model_name.lower()][table_name.lower()]:` line should be `table_name.lower() in MERGING_OPTIMIZATION_FIELDS[model_name.lower()].keys():`

File: specifyweb/specify/views.py [hunk 1]
```diff
@@ -584,9 +584,9 @@ def record_merge_fx(model_name: str, old_model_ids: List[int], new_model_id: int
         # Handle case of updating a large amount of record ids in a foreign table.
         # Example: handle case of updating a large amount of agent ids in the audit logs.
         # Fix by optimizing the query by consolidating it here
-        if model_name.lower() in MERGING_OPTIMIZATION_TABLES.keys() and \
-            table_name.lower() in MERGING_OPTIMIZATION_TABLES[model_name.lower()]:
-            for field_name in MERGING_OPTIMIZATION_FIELDS[(model_name.lower(), table_name.lower())]:
+        if model_name.lower() in MERGING_OPTIMIZATION_FIELDS and \
+            table_name.lower() in MERGING_OPTIMIZATION_FIELDS[model_name.lower()][table_name.lower()]:
```
Comment (acwhite211) [inline specifyweb/specify/views.py]: Should be  `MERGING_OPTIMIZATION_FIELDS.keys()` and `table_name.lower() in MERGING_OPTIMIZATION_FIELDS[model_name.lower()].keys():`

Comment (realVinayak) [reply specifyweb/specify/views.py]: aren't they the same?

x = {'a': 90, 'b': 10}
'a' in x
-- True
 'xx' in x
-- False

Comment (acwhite211) [reply specifyweb/specify/views.py]: ya, tested it out, guess they are the same.  Allows thought you need '.keys()'.  Time to through out all the keys, lol.

**Review by acwhite211 — APPROVED** [review]

File: specifyweb/specify/views.py [hunk 2]
```diff
@@ -458,15 +458,15 @@ def resolve_record_merge_response(start_function, silent=True):
 # Only use if it can be assured that no constraints will be raised, requiring recursive merging.
 MERGING_OPTIMIZATION_TABLES = {
     'agent': {'spauditlog'},
-    'taxon': {'agent'},
-    'collectionobject' : {'agent'}
+    'agent': {'taxon'},
+    'agent' : {'collectionobject'}
```
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py]: are you allowed to have two keys with the same value in python?

Comment (CarolineDenis) [reply specifyweb/specify/views.py]: I think that's an old commit

**Review by realVinayak — APPROVED** [review]
I think this should be used by default -- bulk updating references. If we have a consistent spec where we can store the business rules (which will include uniqueness rules), we can perform bulk update for all columns not having business rules. Some ideas for the future.

**Review by maxxxxxdlp — APPROVED** [review]

**Review by melton-jason — APPROVED** [review]

**Review by bronwyncombs — COMMENTED** [review]
On `fwri_6_1_23`: it seems there is little to no difference in the time it takes for a merge to fail across branches. 

**`v7.9.0`:**

https://github.com/specify/specify7/assets/135047322/e9bcf1ed-5243-4b15-9995-7d749b251f67

[Merging 7beec338-4a5a-4bd6-bc20-079e907bd14f Crash Report - 2023-10-25T16_04_07.321Z.txt](https://github.com/specify/specify7/files/13169757/Merging.7beec338-4a5a-4bd6-bc20-079e907bd14f.Crash.Report.-.2023-10-25T16_04_07.321Z.txt)


**`issue-4019`:**

https://github.com/specify/specify7/assets/135047322/19fe7336-9aed-45cf-afe3-f5d8c56c824f


[Merging 7beec338-4a5a-4bd6-bc20-079e907bd14f Crash Report - 2023-10-25T16_04_07.321Z.txt](https://github.com/specify/specify7/files/13169769/Merging.7beec338-4a5a-4bd6-bc20-079e907bd14f.Crash.Report.-.2023-10-25T16_04_07.321Z.txt)

**Comment by realVinayak** [comment]
Can you use click on 'Linked Records' to see how many records are associated with them? In this case, it is not enough to cause a measurable impact. To test this, use agents which have a lot of records associated with them.

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/specify-7-9-2-release-announcement/1452/1

**Pull request merged**